### ISSUE TITLE
Provider packages should be uninstalled before their replacement gets installed

### ISSUE THREAD

**Issue by Seldaek** [composer/composer#1634]
> - scipy 0.11.0  depends on numpy
> - numpy 1.7.0 is available. Depends on MKL 11.0.0
> - numpy 1.7.0 is also provided by nomkl_numpy 1.7.0, which does not
>   depend on mkl.
> 
> Running the script, I get:
> 
> operations are ...
> Installing mkl (11.0.0)
> Installing numpy (1.7.0)
> Installing scipy (0.11.0)
> Uninstalling nomkl_numpy (1.7.0)
> 
> I would have expected at least:
> 
> operations are ...
> Uninstalling nomkl_numpy (1.7.0)
> Installing mkl (11.0.0)
> Installing numpy (1.7.0)
> Installing scipy (0.11.0)

**Comment by FredM** [comment]
I ran into a related issue recently.
This behaviour (uninstalling before installing) would allow to use both provide and conflict directives. This way you could have two packages providing a virtual package but also conflicting with it (making the first package unique).

Presently it ends up with

```bash
Loading composer repositories with package information
Updating dependencies (including require-dev)
  - Installing my/package-a (1.0)
    Cloning 1.0

  [RuntimeException]
  Failed to execute git clone --no-checkout '' 'vendor/my/package-a' && cd 'vendor/my/package-a' && git remote add composer '' && git fetch composer
  fatal: le dépôt '' n'existe pas
```

Seems like the 'conflict' instruction cleans up its path.
First removing the previously install package corrects this behaviour

A idea if this is planned to be implemented?

**Comment by Seldaek** [comment]
@FredM no concrete plans for this yet since it's possibly complex and kind of an edge case fix.

**Comment by Seldaek** [comment]
I'd say this is fixed in 2.0.
